### PR TITLE
Dependency configuration improvements

### PR DESCRIPTION
**Pull request by ChristopherKoellner**
- [x] make global module testing configuration optional per test/steps level and deprecate
- [x] make type lookup deterministic
- [x] make `provide` force precise type matching

### CODE REVIEW

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/ConfigurationsTestContext.kt [hunk 1]
```diff
@@ -2,16 +2,18 @@ package com.mysugr.sweetest.framework.context
...
 
-internal class ConfigurationsTestContext(private val factories: FactoriesTestContext) {
+class ConfigurationsTestContext(private val factories: FactoriesTestContext) {
 
     private val moduleConfigurations = mutableListOf<ModuleTestingConfiguration>()
 
     val all = moduleConfigurations as List<ModuleTestingConfiguration>
 
-    fun put(configuration: ModuleTestingConfiguration) {
+    fun put(configuration: ModuleTestingConfiguration?) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/ConfigurationsTestContext.kt:11]: hm, somehow it seems to me that it makes no real sense to pass `null` here. can't we just avoid the call if it's null? 🤔

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 2]
```diff
@@ -83,7 +99,35 @@ class DependenciesTestContext {
...
+            }
+
+            block(dependencyState, dependencyState.configuration.defaultDependencyMode)
+        }
+    }
+
+    /**
+     *
+     */
+    private fun assertMode(clazz: KClass<*>, configurationMode: DependencyMode?, mode: DependencyMode) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:128]: this checks for the mode change compared to the configuration but not whether the mode has changed from the last time somebody has enforced a mode. i'll write a test for it and see what it does...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt [hunk 3]
```diff
@@ -20,6 +20,10 @@ class DependencyManager(setupHandlerReceiver: (DependencySetupHandler) -> Unit)
     private val initializerContext = object : DependencyInitializerContext() {
         override fun <T : Any> instanceOf(clazz: KClass<T>): T {
             val configuration = configurations.getAssignableTo(clazz)
+                ?: throw DependencyConfigurations.NotFoundException(
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt:23]: why is that necessary? 🤔

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 2 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:128]: If the mode is set repeatedly the it throws either way. 
```
          if (value != modeField) {
                if (modeField != null) {
                    throw IllegalStateException(
                        "Can't change dependency mode or \"${configuration.clazz}\", it " +
                            "has already been set before and can't be changed afterwards"
                    )
                }
```

File: sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DependenciesAccessor.kt [hunk 4]
```diff
@@ -27,4 +27,8 @@ class DependenciesAccessor(@PublishedApi internal val parent: BaseAccessor) {
     inline fun <reified T : Any> offerMockRequired(noinline initializer: DependencyInitializer<T>) {
         parent.testContext.dependencies.offerMockRequired(T::class, initializer)
     }
+
+    inline fun <reified T : Any> provide() {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DependenciesAccessor.kt:31]: i'd not add it here but rather remove this class as none of it is used anywhere. you can inline the usage of this function, imo

File: sweetest/src/main/java/com/mysugr/sweetest/framework/base/BaseJUnitTest.kt [hunk 5]
```diff
@@ -4,7 +4,12 @@ import com.mysugr.sweetest.framework.build.TestBuilder
 import com.mysugr.sweetest.framework.configuration.ModuleTestingConfiguration
 import org.junit.Before
 
-abstract class BaseJUnitTest(private val moduleTestingConfiguration: ModuleTestingConfiguration) : TestingAccessor {
+abstract class BaseJUnitTest @Deprecated(
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/base/BaseJUnitTest.kt:7]: note to myself: i need to check how this deprecation displays in the IDE when using the lib...

Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/base/BaseJUnitTest.kt:7]: very handy 👌

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt [hunk 6]
```diff
@@ -31,6 +31,8 @@ fun moduleTestingConfiguration(
     val scope = Dsl.MainScope()
     run?.invoke(scope)
 
+    check(scope.dependencies.isNotEmpty()) { "The module configuration must not be empty!" }
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:34]: what is the thought behind this? just thinking: what if someone has just created an empty testing configuration to use it later or there is an empty configuration lying around already? unlikely and edgy but an unnecessary bummer, probably, and strictly speaking a breaking change...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt [hunk 7]
```diff
@@ -74,8 +76,12 @@ class Dsl {
         inline fun <reified T1 : Steps, reified T2 : Steps, reified T3 : Steps, reified R : Any> factory(
             noinline createObject: (T1, T2, T3) -> R
         ) {
-            factories.add(FactoryRunner3(R::class.java, T1::class.java, T2::class.java, T3::class.java,
-                createObject))
+            factories.add(
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:79]: it's time to actually get rid of these 😞  soon then...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyConfigurations.kt [hunk 8]
```diff
@@ -6,7 +6,7 @@ import kotlin.reflect.KClass
 interface DependencyConfigurationConsumer {
     val all: Collection<DependencyConfiguration<*>>
     operator fun <T : Any> get(clazz: KClass<T>): DependencyConfiguration<T>
-    fun <T : Any> getAssignableTo(clazz: KClass<T>): DependencyConfiguration<T>
+    fun <T : Any> getAssignableTo(clazz: KClass<T>): DependencyConfiguration<T>?
     fun <T : Any> getAssignableFrom(clazz: KClass<T>): DependencyConfiguration<T>
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyConfigurations.kt:10]: this function just got inconsistent with `getAssignableTo` (now nullable result), but it's not used. can you remove it?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/ConfigurationsTestContext.kt [hunk 9]
```diff
@@ -2,7 +2,7 @@ package com.mysugr.sweetest.framework.context
 
 import com.mysugr.sweetest.framework.configuration.ModuleTestingConfiguration
 
-internal class ConfigurationsTestContext(private val factories: FactoriesTestContext) {
+class ConfigurationsTestContext(private val factories: FactoriesTestContext) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/ConfigurationsTestContext.kt:5]: why `internal` removed here? as far as i can see you can still compile it when it's here.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 10]
```diff
@@ -1,58 +1,74 @@
...
 
+import com.mysugr.sweetest.framework.dependency.DependencyConfiguration
 import com.mysugr.sweetest.framework.dependency.DependencyInitializer
+import com.mysugr.sweetest.framework.dependency.DependencyManager
 import com.mysugr.sweetest.framework.dependency.DependencyMode
+import com.mysugr.sweetest.framework.dependency.DependencyState
 import com.mysugr.sweetest.framework.environment.TestEnvironment
 import kotlin.reflect.KClass
 
 class DependenciesTestContext {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:11]: note to myself: just checking whether `internal` on this class still works...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 11]
```diff
@@ -83,7 +99,46 @@ class DependenciesTestContext {
...
+    }
+
+    /**
+     * Asserts if the configured mode matches the requested one.
+     */
+    private fun assertMode(clazz: KClass<*>, configurationMode: DependencyMode?, mode: DependencyMode) {
+        check(configurationMode == mode || configurationMode == null) {
+            "Class $clazz has been configured as ${configurationMode?.name} but is requested as ${mode.name}!"
+        }
     }
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:143]: i think you can put the comments into the function name and remove the comments

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 12]
```diff
@@ -83,7 +99,46 @@ class DependenciesTestContext {
...
+                "has already been set before and can't be changed afterwards"
+        }
+    }
+
+    /**
+     * Asserts if the configured mode matches the requested one.
+     */
+    private fun assertMode(clazz: KClass<*>, configurationMode: DependencyMode?, mode: DependencyMode) {
+        check(configurationMode == mode || configurationMode == null) {
+            "Class $clazz has been configured as ${configurationMode?.name} but is requested as ${mode.name}!"
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:141]: for the user it's most important to not only know what exactly has been gone wrong but also what he can do. can you give some hints in "sweetest style"? 😁 
also i think we should not talk about a `Class` (...has been configured) but rather a `Dependency`. it's stated like that mostly throughout the error messages.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 13]
```diff
@@ -83,7 +99,46 @@ class DependenciesTestContext {
...
+
+            block(dependencyState, dependencyState.configuration.defaultDependencyMode)
+        }
+    }
+
+    /**
+     * Asserts if the type [clazz] is already added.
+     */
+    private fun assertDependency(clazz: KClass<*>) {
+        check(TestEnvironment.dependencies.states[clazz] == null) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:130]: when working on improving errors produced by sweetest we should come up with a proper exception type hierarchy in future, but out-of-scope for now...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 14]
```diff
@@ -83,7 +99,46 @@ class DependenciesTestContext {
...
+            "Can't change dependency mode or ${clazz.simpleName}, it " +
+                "has already been set before and can't be changed afterwards"
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:132]: i don't understand the "has already been set before" part. the check checks whether the state is null. so it can't be have been set already, or maybe i miss something?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt [hunk 15]
```diff
@@ -19,7 +19,7 @@ class DependencyManager(setupHandlerReceiver: (DependencySetupHandler) -> Unit)
 
     private val initializerContext = object : DependencyInitializerContext() {
         override fun <T : Any> instanceOf(clazz: KClass<T>): T {
-            val configuration = configurations.getAssignableTo(clazz)
+            val configuration = requireNotNull(configurations.getAssignableTo(clazz))
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt:22]: now that `getAssignableTo` doesn't throw an exception with a well explaining error message anymore it should be done here i think

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 16]
```diff
@@ -83,7 +99,46 @@ class DependenciesTestContext {
...
+            }
+
+            block(dependencyState, dependencyState.configuration.defaultDependencyMode)
+        }
+    }
+
+    /**
+     * Asserts if the type [clazz] is already added.
+     */
+    private fun assertDependency(clazz: KClass<*>) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:129]: this check causes a lot of tests to fail in production when i tried it. we should further investigate this and add tests in sweetest for those circumstances...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencySetup.kt [hunk 17]
```diff
@@ -28,28 +28,24 @@ object DependencySetup {
     inline fun <reified T : Any> addRealRequired(
         alias: KClass<*>? = null,
         noinline realInitializer: DependencyInitializer<T>
-    ):
-        DependencyConfiguration<T> =
-        setupHandler.addConfiguration(
-            T::class, realInitializer, null,
-            DependencyMode.REAL, alias)
+    ): DependencyConfiguration<T> =
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencySetup.kt:31]: i wonder about this auto-format. was the column count not set correctly? 🤔 
did you do it with ktlint's auto-formatter? maybe it doesn't pick up the correct max line length...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencySetup.kt [hunk 18]
```diff
@@ -28,28 +28,24 @@ object DependencySetup {
...
+        setupHandler.addConfiguration(T::class, realInitializer, null, DependencyMode.REAL, alias)
 
     inline fun <reified T : Any> addMocked(
         alias: KClass<*>? = null,
         noinline mockInitializer: DependencyInitializer<T>
-    ):
-        DependencyConfiguration<T> =
+    ): DependencyConfiguration<T> =
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencySetup.kt:37]: way better here though 😬

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyStates.kt [hunk 19]
```diff
@@ -25,8 +26,7 @@ class DependencyStates(private val initializerContext: DependencyInitializerCont
 
     override val all get() = statesMap.values
 
-    override operator fun <T : Any> get(configuration: DependencyConfiguration<T>):
-        DependencyState<T> {
+    override operator fun <T : Any> get(configuration: DependencyConfiguration<T>): DependencyState<T> {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyStates.kt:29]: instead of overriding the index operator i we should have functions `getByConfiguration` and `getByDependencyType` instead of the index operator, i think. as it is now, it's hard to understand what's happening from the caller's side...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt (hunk 3 above)
Comment (eder7) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt]: i think my question caused you removing this. but it was just meant as a question. i think that should be reintroduced. sorry for the confusion 😬

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt [hunk 20]
```diff
@@ -20,6 +20,10 @@ class DependencyManager(setupHandlerReceiver: (DependencySetupHandler) -> Unit)
     private val initializerContext = object : DependencyInitializerContext() {
         override fun <T : Any> instanceOf(clazz: KClass<T>): T {
             val configuration = configurations.getAssignableTo(clazz)
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt]: as talked about f2f: please

* extract and unify behavior of `instanceOf` with `by dependency` 
* remove the auto-provision when neither a configuration nor state was found for the type requested

thx 🙂

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 10 above)
Comment (eder7) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:11]: yes, works, but you need to put `@PublishedApi` on it. maybe you can unify that in the `*Context` classes?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt (hunk 6 above)
Comment (eder7) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:34]: fails a test currently 😬

Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:34]: What would be the purpose of creating an empty testing configuration? And if there are empty ones lying around they should be cleaned up IMO.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DelegatesAccessor.kt [hunk 21]
```diff
@@ -15,23 +15,37 @@ class DelegatesAccessor(@PublishedApi internal val accessor: BaseAccessor) {
...
-            val dependency = TestEnvironment.dependencies.configurations.getAssignableTo(T::class)
-            val dependencyState = TestEnvironment.dependencies.states[dependency]
+
+            val dependencyState = TestEnvironment.dependencies.configurations.getAssignableTo(T::class)?.let {
+                TestEnvironment.dependencies.states[it]
+            } ?: kotlin.run {
+                TestEnvironment.dependencies.states[T::class]
+            } ?: kotlin.run {
+                accessor.dependencies.provide<T>()
+                requireNotNull(TestEnvironment.dependencies.states[T::class])
```
[the remark this answers is not in the document]
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DelegatesAccessor.kt:40]: Removed this fallback as discussed.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt (hunk 6 above)
Comment (eder7) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:34]: as talked about f2f: replace by a `println` deprecation message (pointing towards sweetest 2.0.0 probably...?)

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 11 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:143]: I don't really get it. What do you mean?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 14 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:132]: Hm you are right. That makes no sense. When I think about it, the whole check is redundant as we should just ignore multiple dependency adds.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DelegatesAccessor.kt [hunk 22]
```diff
@@ -32,13 +32,13 @@ class DelegatesAccessor(@PublishedApi internal val accessor: BaseAccessor) {
...
-                TestEnvironment.dependencies.states[T::class]
-            } ?: kotlin.run {
-                accessor.dependencies.provide<T>()
-                requireNotNull(TestEnvironment.dependencies.states[T::class])
-            }
+                TestEnvironment.dependencies.states.getByConfiguration(it)
+            } ?: run {
+                TestEnvironment.dependencies.states.getByDependencyType(T::class)
+            } ?: error(
+                "No configuration or dependency state for ${T::class.simpleName} added. " +
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DelegatesAccessor.kt:39]: can you write an example what that means? e.g. "Add the dependency with "provide<$T::class.simpleName>" (or the deprecated way: to the module configuration)"

File: sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DependenciesAccessor.kt [hunk 23]
```diff
@@ -27,8 +27,4 @@ class DependenciesAccessor(@PublishedApi internal val parent: BaseAccessor) {
     inline fun <reified T : Any> offerMockRequired(noinline initializer: DependencyInitializer<T>) {
         parent.testContext.dependencies.offerMockRequired(T::class, initializer)
     }
-
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/accessor/DependenciesAccessor.kt]: i think you can delete that class entirely

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 24]
```diff
@@ -13,8 +13,10 @@ class DependenciesTestContext {
     @Deprecated("Use \"provide\" instead.", replaceWith = ReplaceWith("provide"))
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:13]: this class could be made internal i think...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 25]
```diff
@@ -13,8 +13,10 @@ class DependenciesTestContext {
...
     fun requireReal(clazz: KClass<*>) {
         val mode = DependencyMode.REAL
-        addDependency(clazz) { state, configurationMode ->
-            assertMode(clazz, configurationMode, mode)
+        assertDependencyMode(clazz, mode)
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:16]: what about `checkModeChangePossible` (it's not really an assertion i think)?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 26]
```diff
@@ -112,33 +121,34 @@ class DependenciesTestContext {
      */
     private fun addDependency(clazz: KClass<*>, block: (DependencyState<out Any>, DependencyMode?) -> Unit) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:122]: what about `prepareAndUseDependencyOf`?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 27]
```diff
@@ -13,8 +13,10 @@ class DependenciesTestContext {
     @Deprecated("Use \"provide\" instead.", replaceWith = ReplaceWith("provide"))
     fun requireReal(clazz: KClass<*>) {
         val mode = DependencyMode.REAL
-        addDependency(clazz) { state, configurationMode ->
-            assertMode(clazz, configurationMode, mode)
+        assertDependencyMode(clazz, mode)
+        addDependency(clazz) { state, _ ->
+            // The assertion is omitted to keep compatibility to older versions of this library.
+            // assertConfiguredMode(clazz, configurationMode, mode)
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:19]: ❓

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 28]
```diff
@@ -112,33 +121,34 @@ class DependenciesTestContext {
...
-     * Asserts if the type [clazz] is already added.
+     * Asserts if the configured mode matches the requested one.
      */
-    private fun assertDependency(clazz: KClass<*>) {
-        check(TestEnvironment.dependencies.states[clazz] == null) {
-            "Can't change dependency mode or ${clazz.simpleName}, it " +
-                "has already been set before and can't be changed afterwards"
+    private fun assertConfiguredMode(clazz: KClass<*>, configurationMode: DependencyMode?, mode: DependencyMode) {
+        check(configurationMode == mode || configurationMode == null) {
+            "Dependency $clazz has been configured as ${configurationMode?.name} but is requested as ${mode.name}! "
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:139]: i think this message can still be a bit more conclusive. what about:

```suggestion
            "Dependency \"$clazz.simpleName\" has been forced to be ${configurationMode?.name} in the module testing configuration, but you requested it to be ${mode.name} instead. Please loosen up the constraint by using \"any\" instead of \"requireMock\" or \"requireReal\" or remove the module testing configuration entirely as it's deprecated anyway."
```

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 29]
```diff
@@ -112,33 +121,34 @@ class DependenciesTestContext {
...
         }
     }
 
     /**
-     * Asserts if the configured mode matches the requested one.
+     * Asserts if the [DependencyState.mode] of an already added [DependencyState] has changed.
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:144]:
```suggestion
     * Checks if the [DependencyState.mode] of an already added [DependencyState] is requested to be changed, which is not allowed.
```

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 30]
```diff
@@ -112,33 +121,34 @@ class DependenciesTestContext {
...
      */
-    private fun assertMode(clazz: KClass<*>, configurationMode: DependencyMode?, mode: DependencyMode) {
-        check(configurationMode == mode || configurationMode == null) {
-            "Class $clazz has been configured as ${configurationMode?.name} but is requested as ${mode.name}!"
+    private fun assertDependencyMode(clazz: KClass<*>, mode: DependencyMode) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:146]:
```suggestion
    private fun checkDependencyMode(clazz: KClass<*>, mode: DependencyMode) {
```

File: sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt [hunk 31]
```diff
@@ -19,9 +21,30 @@ class DependencyManager(setupHandlerReceiver: (DependencySetupHandler) -> Unit)
 
     private val initializerContext = object : DependencyInitializerContext() {
         override fun <T : Any> instanceOf(clazz: KClass<T>): T {
-            val configuration = requireNotNull(configurations.getAssignableTo(clazz))
-            val state = states[configuration]
-            return state.instance
+            try {
+                if (clazz.isSubclassOf(BaseSteps::class)) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/dependency/DependencyManager.kt]: good you thought about it 👍

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 27 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:19]: As discussed.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt (hunk 24 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:13]: it could when adding @PublishedApi to every method. They are accessed form the `BaseBuilder`.

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 32]
```diff
@@ -1,89 +1,145 @@
...
 
+import com.mysugr.sweetest.framework.base.SweetestException
 import com.mysugr.sweetest.framework.dependency.DependencyInitializer
 import com.mysugr.sweetest.framework.dependency.DependencyMode
 import com.mysugr.sweetest.framework.environment.TestEnvironment
 import kotlin.reflect.KClass
 
 class DependenciesTestContext {
 
-    @Deprecated("Use \"provide\" instead.", replaceWith = ReplaceWith("provide"))
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:10]: no public API, so not needed

File: sweetest/src/test/java/com/mysugr/sweetest/DependenciesTest.kt [hunk 33]
```diff
@@ -114,10 +144,141 @@ class DependenciesTest {
         }
     }
 
+    @Test(expected = SweetestException::class)
+    fun `Given module dependency configuration, changing REAL to MOCK throws no exception`() {
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/DependenciesTest.kt:148]: As far as I can see this should also throw, shouldn't it?

File: sweetest/src/test/java/com/mysugr/sweetest/DependenciesTest.kt [hunk 34]
```diff
@@ -131,4 +292,11 @@ class DependenciesTest {
...
         }
     }
+
+    private fun givenAllAny() {
+        moduleTestingConfiguration = moduleTestingConfiguration {
+            dependency any of<AService>()
+            dependency any of<BViewModel>()
+        }
+    }
 }
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/DependenciesTest.kt:302]: new line missing

File: sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt [hunk 35]
```diff
@@ -0,0 +1,86 @@
...
+
+            override fun configure() = super.configure()
+                .requireReal<Cat>() // already fails here
+        }
+
+        test.junitBefore()
+
+        test.cat
+    }
+}
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt:86]: newline

File: sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt [hunk 36]
```diff
@@ -0,0 +1,86 @@
...
+    @After
+    fun tearDown() {
+        TestEnvironment.fullReset()
+    }
+
+    @Test
+    fun `Consuming supertype gives subtype`() {
+
+        val config = moduleTestingConfiguration {
+            dependency realOnly of<Cat>()
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt:25]: what would happen when we have another class with parent Animal configured here. 
If there is no test for it yet it would be nice to have one.

File: sweetest/src/test/java/com/mysugr/sweetest/PolymorphismTest.kt [hunk 37]
```diff
@@ -0,0 +1,198 @@
...
+            val animal by dependency<Animal>() // matches only Animal
+
+            override fun configure() = super.configure()
+                .provide<Cat>()
+                .provide<Animal> { mock() }
+        }
+
+        assertNotSame(test.cat, test.animal) // precise matching --> different types
+    }
+}
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/PolymorphismTest.kt:198]: newline

**Review by ChristopherKoellner — COMMENTED** [review]
All in all it looks solid to me. 
I just have a few nitpicks however, nothing major. 
Have you already tried to run it in the shared or logbook repository? 🤔

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt [hunk 38]
```diff
@@ -31,6 +32,10 @@ fun moduleTestingConfiguration(
     val scope = Dsl.MainScope()
     run?.invoke(scope)
 
+    if (scope.dependencies.isEmpty()) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt:35]: is it ok to remove this? i think the deprecation of the constructor taking the module testing configuration should be enough? 🤔

File: sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt (hunk 36 above)
Comment (eder7) [reply sweetest/src/test/java/com/mysugr/sweetest/LegacyPolymorphismTest.kt:25]: added a test, like it?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt (hunk 38 above)
Comment (ChristopherKoellner) [reply sweetest/src/main/java/com/mysugr/sweetest/framework/configuration/ModuleTestingConfiguration.kt]: Fine for me

**Review by ChristopherKoellner — COMMENTED** [review]
I would approve it, if it wouldn't be my PR. 😛 
Thanks for adding the additional testcase and also for creating more meaningful error messages. 
LGTM.

**Comment by eder7** [comment]
Thanks for cross-checking @ChristopherKoellner 🙂 
Checked compatibility with our code (both repositories) all working (also on CI) ✅
